Docs/gh 6 read api spec

jmchamica opened the pull request:
The public documentation goes over the representations of the different API resources.
The structure is as follows:
- properties
  - type, examples
- link relations
  - rel
- actions
  - fields
  - safety
  - if they are templated
- example representations

Closes #6

Commit:
First Event reade API design.

Commit:
Revision. Changed from endpoint oriented to class oriented.

Commit:
Added forgotten about link relations.

jmchamica committed:
Documentation for the read API's Courses and Class Sections endpoints.

The documents go over: the representation's structure, properties, link
relations, query parameters and actions.

jmchamica committed:
Adds Root resource documentation (read API)

alexandru-pena committed:
Adding class read api documentation

JorgePBrown committed:
Merge pull request #7 from i-on-project/docs/gh-6-read-api-events

Docs: Read API Spec - Events

jmchamica committed:
Use the same document structure, for consistency.

DSantos44774 commented on docs/api/courses.md:
```diff
@@ -0,0 +1,153 @@
+## Properties
+* `acronym`: the course's unique acronym; an abbreviation of its name
+  - type: **text**
+  - e.g. "DWA"
```
Shouldn't it be "DAW" instead of "DWA"?

jmchamica replied:
The name of the course is "Development of Web Apps". Therefore, the acronym should be "DWA".

JorgePBrown commented on docs/api/events.md:
```diff
@@ -0,0 +1,325 @@
+
+## Link Relations:
+An event representation:
+* *must* include a link to its context, using the `self` link relation
```
@joao-pereira5 The context is supposed to be linked to using the `about` relation right?

jmchamica replied:
Cited from the IANA registry: 
`about` rel: `Refers to a resource that is the subject of the link's context.`
`self` rel: `Conveys an identifier for the link's context.`

JorgePBrown commented on docs/api/events.md:
```diff
@@ -0,0 +1,325 @@
+  - `title`: title of the `Period`
+
+## Link Relations:
+* `self`: the context of the schedule
```
Same thing about context.

JorgePBrown commented on docs/api/events.md:
```diff
@@ -0,0 +1,325 @@
+
+## Link Relations:
+A task representation:
+* *must* include a link to its context, using the `self` link relation
```
Same thing about context.

jmchamica committed:
Corrections of the read API documentation.

pmhsfelix commented on docs/api/class.md:
```diff
@@ -0,0 +1,110 @@
+# `Class`
+
+The `Class` endpoint returns all classes of all calendar terms of given `Course`, the following example returns all classes of course "WAD". The full representation of a `Class`, in the [application/vnd.siren+json](https://github.com/kevinswiber/siren) media type.
```
We are not documenting "endpoints". We are documenting representations.

pmhsfelix replied:
Adds the "class" value.

pmhsfelix commented on docs/api/class.md:
```diff
@@ -0,0 +1,110 @@
+  - type: **text**
+  - e.g. "WAD"
+
+* `coordinator`: the `lecturer` responsible for managing the `class`
```
We may not have this in R1.

pmhsfelix commented on docs/api/class.md:
```diff
@@ -0,0 +1,110 @@
+  - type: **text**
+  - e.g. "WAD"
+
+* `mandatory`:
```
May not be available in R1.

pmhsfelix commented on docs/api/class.md:
```diff
@@ -0,0 +1,110 @@
+  - e.g. true
+
+* `credits`:
+  - type: **number**
```
Same

pmhsfelix commented on docs/api/class.md:
```diff
@@ -0,0 +1,110 @@
+  - e.g. 6
+
+* `hours of work`:
+  - type: **number**
```
Same

pmhsfelix commented on docs/api/class.md:
```diff
@@ -0,0 +1,110 @@
+  "actions": [],
+  "links": [
+    { "rel" : [ "self" ], "href": "/v0/courses/wad/classes/1920v" },
+    { "rel" : [ "class" ], "href": "/v0/courses/wad/classes/" },
```
Probably the _term_ is also a related resource.

pmhsfelix commented on docs/api/class.md:
```diff
@@ -0,0 +1,110 @@
+  "class" : [ "collection", "class" ],
+  "properties": {
+    "course": "WAD",
+    "size": 5
```
May remove

pmhsfelix commented on docs/api/events.md:
```diff
@@ -0,0 +1,336 @@
+  - type: **text**
+  - e.g. "First exam for the WAD-s1920v class. Students are free to bring a cheat sheet of only 1 page, written by hand."
+
+* `start_date`: starting date of the `Event`
```
Be consistent on the identifier style (snake_case vs camelCase).

pmhsfelix commented on docs/api/events.md:
```diff
@@ -0,0 +1,336 @@
+
+* `start_date`: starting date of the `Event`
+  - type: **date**
+  - e.g. 19/03/2020
```
Use the ISO data format. Local time with offset.

pmhsfelix commented on docs/api/events.md:
```diff
@@ -0,0 +1,336 @@
+# `Event`
```
Use iCalendar concepts

Commit:
Time Standard and naming consistency changes

Changed date and time to ISO8601 format and all naming to camelCase

Commit:
Updated `Event` segment of the API. Missing Todo and Journal.

jmchamica commented on docs/api/events.md:
```diff
@@ -0,0 +1,520 @@
+All types are iCalendar([RFC5545](https://tools.ietf.org/html/rfc5545) and [RFC7986](https://tools.ietf.org/html/rfc7986)) types: `Event`; `Journal`; `Todo`.
+iCalendar and [jCal](https://tools.ietf.org/html/rfc7265) formats are supported and can be obtained by using `text/calendar` or `application/calendar+json`, respectively, in the `Accept` header of requests. If the request made has `application/vdn.siren+json` on the `Accept` header then a in-house representation will be used. This representation still holds the same information as jCal, it simply ...
+
+iCalendar components support sub-components, however, the components being used(`Event`, `Journal` and `Todo`) cannot be nested in other components so there would never exist any subcomponents. As a consequence, the `subComponents` property of the objects has been ommitted.
```
The first sentence is confusing. What does it mean? :thinking:

jmchamica commented on docs/api/events.md:
```diff
@@ -0,0 +1,520 @@
+Date formatting is done according to [RFC3339](https://tools.ietf.org/html/rfc3339) which is an Internet profile of the [ISO8601](https://www.iso.org/iso-8601-date-and-time-format.html).
+
+# `Event`
+An `Event`, as described [here](https://tools.ietf.org/html/rfc5545#section-3.6.1) describes an occurrence within a given time frame, such as an exam, a meeting or an appointment. It can also represent a repeating occurrence if the [`rrule`](https://tools.ietf.org/html/rfc5545#section-3.8.5.3) property is specified. __Not all properties mentioned in the RFC are supported by this API, only the one ...
```
> as described here describes

I suggest changing to something along the lines of: "An Event represents/is(<- link here) an occ...

JorgePBrown replied:
I have "an Event, described here, represents [...]" in my local branch. Is that ok?

joao-pereira5 commented on docs/api/events.md:
````diff
@@ -0,0 +1,520 @@
+}
+```
+
+# `Schedule`
````
Should we use Schedule instead of Recurrence Rule? Either one is fine for me, but using both simultaneously would be very inconsistent.

joao-pereira5 commented on docs/api/events.md:
```diff
@@ -0,0 +1,520 @@
+
+# `Todo`
+
+An `Event` that lacks a `startDate` and is coupled with some sort of delivery. For example, a work assignment at school.
```
A todo's start date could be its creation date.

jmchamica reviewed (commented):
Some minor suggestions/observations.

jmchamica committed:
Corrections to the read API, and addition of more actions.

Removed some of the `Class` resource's properties, leaving only the
essential ones.
Actions for searching, adding and deleting items from a Class collection
were added. Actions for editing and removing a Class were added.

Commit:
Changes to all of the Event read api according to iCalendar

Commit:
Merge branch 'docs/gh-6-read-api-spec' of https://github.com/i-on-project/core into docs/gh-6-read-api-spec

jmchamica committed:
Added write actions in the Course and Event representation.

jmchamica committed:
Added Term representation. Fixed a few errors in others.

jmchamica committed:
Fixes to the API spec. Additional links added.
